File: principles-of-programming-languages/assignment-2/assignment-2/PrinciplesOfProgrammingLanguagesAssignment2/src/BasamakKontrol.c
// @ Rümeysa Ceran

#include "NesneleriOlustur.h"
#include "Ilce.h"
#include "Sehir.h"
#include "Mahalle.h"
#include "Kisi.h"
#include "BasamakKontrol.h"

BasamakKontrol BasamakKontrolOlusturma(){
    BasamakKontrol this = (BasamakKontrol)malloc(sizeof(struct BASAMAK_KONTROL));

    this->basamakKontroluIlce      = (Ilce* (*)(struct BASAMAK_KONTROL*, Sehir, NesneleriOlustur))&BasamakKontroluIlce;
    this->basamakKontroluMahalle   = &BasamakKontroluMahalle;
    this->basamakKontroluKisi      = &BasamakKontroluKisi;
    this->nufusBasamakSayisiKontrolu = (void (*)(struct BASAMAK_KONTROL*, Sehir**, int*, const NesneleriOlustur))&NufusBasamakSayisiKontrolu;
    this->yeniIlceSayisiAta        = (void (*)(struct BASAMAK_KONTROL*, int))&YeniIlceSayisiAta;
    this->yoket                    = (void (*)(struct BASAMAK_KONTROL*))&BasamakKontrolYoket;
    this->yeniIlceSayisi           = 0;

    return this;
}

//ilce sayisi 1 mi yoksa daha fazla mı kontrolu yaparak gerekli islemleri yapar
Ilce* BasamakKontroluIlce(const BasamakKontrol this, Sehir sehir, const NesneleriOlustur nesneleriOlustur){
    int ilce_sayisi     = sehir->ilceSayisi;
    int yeni_ilce_sayisi = ilce_sayisi / 2;

   
    if (yeni_ilce_sayisi != 0){//ilce sayisi 1 den fazla is ilceleri ikiye bolerek yeni ilceler olusturur
        Ilce* yeni_ilceler = (Ilce*)malloc(sizeof(Ilce) * yeni_ilce_sayisi);

        // eski sehirdeki ilceleri yeni_ilceler dizisine atar
        for (int i = 0; i < yeni_ilce_sayisi; i++)
        {       
            yeni_ilceler[i] = sehir->ilceler[i];
        }
        
        //eski sehirde kalan ilce sayısı hesaplanir
        int kalan = ilce_sayisi - yeni_ilce_sayisi;
        //kalan_ilceler dizisine yeni_ilceler dısında kalan ilceler atanir
        Ilce* kalan_ilceler =
        malloc(sizeof(Ilce) * kalan);
        for (int i = 0; i < kalan; i++) {

        kalan_ilceler[i] =
        sehir->ilceler[yeni_ilce_sayisi + i];
        }

        //eski adres bırakılır
        free(sehir->ilceler);

        // kalan ilceler eski sehrin ilceler dizisine atanir
        sehir->ilceler = kalan_ilceler;
        // eski sehrin ilce sayisi guncellenir
        sehir->ilceSayisi = kalan;
        
        //yeni ilce sayısı guncellenir
        this->yeniIlceSayisiAta(this, yeni_ilce_sayisi);

        return yeni_ilceler;
    }
    else{//ilce sayısının 1 oldugu durumda mahalle sayılarını bolmek icin BasamakKontorolMahalle fonksiyonuna gider
        this->yeniIlceSayisiAta(this, 1); // bu durumda olusacak yeni ilce sayısı 1 olur
        return this->basamakKontroluMahalle(this,sehir->ilceler[0], nesneleriOlustur);
    }
}


//mahalle sayisinin 1 den fazla olup olmadıgına bakar ve buna gore mahalle bolme islemini gerceklestirir
Ilce* BasamakKontroluMahalle(const BasamakKontrol this, Ilce ilce, const NesneleriOlustur nesneleriOlustur){
    int yeni_mahalle_sayisi = (ilce->mahalleSayisi)/ 2;

    
    if (yeni_mahalle_sayisi != 0){//mahalle sayısının 1 den fazla oldugu durumdur yani mahalle sayııs ikiye bolunur
        Ilce* yeni_ilceler =(Ilce*)malloc(sizeof(Ilce));
        Mahalle* yeni_mahalleler = (Mahalle*)malloc(sizeof(Mahalle) * yeni_mahalle_sayisi);

        //eski ilcenin mahallelerinden yeni mahalle sayisi kadar mahalle yeni_mahalleler dizisine atanir
        for (int i = 0; i < yeni_mahalle_sayisi; i++)
            yeni_mahalleler[i] = ilce->mahalleler[i];

        //yeni olusacak ilcenin nufusu hesaplanir
        int yeni_nufus = yeni_mahalle_sayisi * yeni_mahalleler[0]->super->nufus;

        //eski ilcenin nufusu hesaplanir
        int kalan = ilce->mahalleSayisi - yeni_mahalle_sayisi;

        //yeni mahallelerin atanma isleminden geriye kalan mahalleler, kalan mahalleler dizisine atanir
        Mahalle* kalan_mahalleler =
            malloc(sizeof(Mahalle) * kalan);

        for (int i = 0; i < kalan; i++) {
            kalan_mahalleler[i] =
               ilce->mahalleler[yeni_mahalle_sayisi + i];
        }

        free(ilce->mahalleler);

        //eski ilceye kalan mahalleler atanir
        ilce->mahalleler = kalan_mahalleler;
        ilce->mahalleSayisi = kalan; //eski ilcenin mahalle sayisi guncellenir
        ilce->super->nufus -= yeni_nufus; // eski ilcenin nufusu guncellenir

        // yeni ilce icin dosyadan isim getirilir
        char* ilce_ad = malloc(100);
        char ilce_isim[1][100];
        nesneleriOlustur->kayitlariGetir(nesneleriOlustur->ilce_dosya, ilce_isim, 1);
        strcpy(ilce_ad, ilce_isim[0]);

        //yeni ilce olusturulur
        Ilce yeni_ilce = IlceOlusturma(ilce_ad, yeni_nufus);
        yeni_ilce->mahalleler   = yeni_mahalleler;
        yeni_ilce->mahalleSayisi = yeni_mahalle_sayisi;
        yeni_ilceler[0] = yeni_ilce;

        return yeni_ilceler;
    }else{
        return this->basamakKontroluKisi(ilce, nesneleriOlustur);
    }
}


Ilce* BasamakKontroluKisi(Ilce ilce, const NesneleriOlustur nesneleriOlustur){
    Mahalle mevcut = ilce->mahalleler[0]; 
    int yeni_kisi_sayisi = (mevcut->super->nufus) / 2;

    Ilce* yeni_ilceler = (Ilce*)malloc(sizeof(Ilce));
    Kisi* yeni_kisiler = (Kisi*)malloc(sizeof(Kisi) * yeni_kisi_sayisi);


    // eski mahalleden yeni kisi sayisi kadar kisi yeni_kisiler dizisine atanir
    for (int i = 0; i < yeni_kisi_sayisi; i++){
         yeni_kisiler[i] = mevcut->kisiler[i];
    }
       
    // eski mahallenin nufusu hesaplanir
    int kalan = mevcut->super->nufus - yeni_kisi_sayisi;

    //yeni kisilerden kalan kisiler kalan_kisiler dizisine atanir
    Kisi* kalan_kisiler =
        malloc(sizeof(Kisi) * kalan);

    for (int i = 0; i < kalan; i++) {
        kalan_kisiler[i] =
            mevcut->kisiler[yeni_kisi_sayisi + i];
    }

    free(mevcut->kisiler);

    // eski mahallenin kisilerine kalan kisler atanir
    mevcut->kisiler = kalan_kisiler;
    // eski mahallenin nufusu guncellenir
    mevcut->super->nufus = kalan;

    //eski ilcenin nufusu hesaplanir
    ilce->super->nufus   -= yeni_kisi_sayisi;

    //mahalle adi dosyadan getirilir
    char* mahalle_ad = malloc(100);
    char mahalle_isim[1][100];
    nesneleriOlustur->kayitlariGetir(nesneleriOlustur->mahalle_dosya, mahalle_isim, 1);
    strcpy(mahalle_ad, mahalle_isim[0]);

    // yeni mahalle olusturulur
    Mahalle yeni_mahalle = MahalleOlusturma(mahalle_ad, yeni_kisi_sayisi);
    yeni_mahalle->kisiler = yeni_kisiler;

    Mahalle* yeni_mahalleler =(Mahalle*)malloc(sizeof(Mahalle));
    yeni_mahalleler[0] = yeni_mahalle;

    //yeni ilce adi dosyadan getirilir
    char* ilce_ad = malloc(100);
    char ilce_isim[1][100];
    nesneleriOlustur->kayitlariGetir(nesneleriOlustur->ilce_dosya, ilce_isim, 1);
    strcpy(ilce_ad, ilce_isim[0]);

    //yeni ilce olusturulur
    Ilce yeni_ilce = IlceOlusturma(ilce_ad, yeni_kisi_sayisi);
    yeni_ilce->mahalleler   = yeni_mahalleler;
    yeni_ilce->mahalleSayisi = 1;
    yeni_ilceler[0] = yeni_ilce;

    return yeni_ilceler;
}


// yeni ilce sayısını degistirmek için kullanılır
void YeniIlceSayisiAta(const BasamakKontrol this, int sayi){
    this->yeniIlceSayisi = sayi;
}


// nufusun 4 basamaklı olup olmama durumuna gore islemler gerceklesir
void NufusBasamakSayisiKontrolu(const BasamakKontrol this, Sehir** sehirler, int* sehirSayisi, const NesneleriOlustur nesneleriOlustur){
    // yeni sehirler icin bellekte alan acılır
    Sehir* yeniSehirler = (Sehir*)malloc(sizeof(Sehir));
    int yeniSayac = 0;

    // tum sehirler sırasıyla nufuslarının basamak sayısına gore kontrol edilir
    for (int i = 0; i < *sehirSayisi; i++){
        if (((*sehirler)[i]->super->nufus) > 999){ //  4 basamaktan fazla ise
            Ilce* yeni_ilceler = this->basamakKontroluIlce(this, (*sehirler)[i], nesneleriOlustur); // once ilceler bolunuyor mu kontrolu

            int yeni_ilce_sayisi = this->yeniIlceSayisi; // yein olusan ilce sayısı alınır
            int yeni_nufus = yeni_ilce_sayisi * yeni_ilceler[0]->super->nufus; // yeni sehrin nufusu hesaplanir

            (*sehirler)[i]->super->nufus -= yeni_nufus; // eski sehrin nufusu azaltırılır

            // sehir adı nesneleri olustur struct'ın fonksiyonu olan kayitlariGetir fonksiyonu ile dosyadan getirilir
            char* sehir_ad = malloc(100);
            char isim[1][100];
            nesneleriOlustur->kayitlariGetir(nesneleriOlustur->sehir_dosya, isim, 1);
            strcpy(sehir_ad, isim[0]);

            //yeni sehir olusturulur
            Sehir yeni_sehir = SehirOlusturma(sehir_ad, yeni_nufus);
            yeni_sehir->ilceler   = yeni_ilceler; // yeni ilceler yeni sehre atanir
            yeni_sehir->ilceSayisi = yeni_ilce_sayisi; // sehrin ilce sayısı atanir


            yeniSehirler = realloc(yeniSehirler, (yeniSayac + 1) * sizeof(Sehir)); //
            yeniSehirler[yeniSayac++] = yeni_sehir; // yeni_sehir en son sehirler dizisine eklenmek uzere  yeniSehirler dizisine eklenir

            this->yeniIlceSayisi = 0; //yeniIlceSayisi diger sehirler icin sıfırlanır
        }
    }

    Sehir* tmp = realloc(*sehirler, (*sehirSayisi + yeniSayac) * sizeof(Sehir));
    if (!tmp) return;

    *sehirler = tmp; // sehriler dizisinin alanı artırılır

    for (int i = 0; i < yeniSayac; i++) // sehirler dizisine yeni sehirler atanir
        (*sehirler)[*sehirSayisi + i] = yeniSehirler[i];

    *sehirSayisi += yeniSayac; // sehirSayisi guncellenir
    free(yeniSehirler); // yeniSehirler dizisini tutan adres bırakılır
}

// yok edici fonskiyon
void BasamakKontrolYoket(const BasamakKontrol this){
    if(this == NULL) return;
    free(this);
}

File: principles-of-programming-languages/assignment-1/assignment-1/PrinciplesOfProgrammingLanguagesAssignment1/src/PrinciplesOfProgrammingLanguagesAssignment1/BasamakKontrol.java
// @ Rümeysa Ceran

package PrinciplesOfProgrammingLanguagesAssignment1;

import java.util.ArrayList;

import com.github.javafaker.Faker;

public class BasamakKontrol {

	public BasamakKontrol(ArrayList<Sehir> sehirler) {
	}

	public static Faker faker = new Faker();

	// sehrin ilce sayısını ikiye böler, eger tek ilçe varsa basamakKontroluMahalle
	// fonksiyonunu döndürür
	public static ArrayList<Ilce> basamakKontroluIlce(Sehir sehir) {
		int ilce_sayisi = sehir.ilceler.size();
		int yeni_ilce_sayisi = ilce_sayisi / 2;
		if (yeni_ilce_sayisi != 0) {// ilce sayısı birden fazla
			ArrayList<Ilce> yeni_ilceler = new ArrayList<>();
			for (int j = 0; j < yeni_ilce_sayisi; j++) { // sehrin ilce sayısının yarısı kadar ilceyi yeni_ilceler
															// listesine ekler
				yeni_ilceler.add(sehir.ilceler.get(j));
			}
			for (int j = yeni_ilce_sayisi - 1; j >= 0; j--) { // sehirden yeni sehre eklenen ilceleri mevcut sehirden
																// siler
				sehir.ilceler.remove(j);
			}
			return yeni_ilceler;
		} else {
			return basamakKontroluMahalle(sehir.ilceler.get(0));
		} // ilce sayısı 1'dir, ilceler bölünemez o yüzden mahalleler bölünmeli

	}

	// mahalle sayısını ikiye böler ve yeni ilce olusturur ve mahallelerin yarısını
	// yeni ilçeye gecirir, eger mahalle sayısı 1 ise basamakSayisiKontroluKisi
	// fonksiyonunu cagırır.
	public static ArrayList<Ilce> basamakKontroluMahalle(Ilce ilce) {
		int yeni_mahalle_sayisi = ilce.mahalleler.size() / 2;
		if (yeni_mahalle_sayisi != 0) { // mahalle sayısı 1'den fazla
			ArrayList<Ilce> yeni_ilceler = new ArrayList<>();
			ArrayList<Mahalle> yeni_mahalleler = new ArrayList<>();
			int yeni_nufus = 0;
			for (int j = 0; j < yeni_mahalle_sayisi; j++) { // ilcenin mahallelerinin yarısı yeni_mahalleler dizisine
															// eklenir
				yeni_mahalleler.add(ilce.mahalleler.get(j));
			}
			yeni_nufus = yeni_mahalle_sayisi * yeni_mahalleler.get(0).nufus; // yeni olusacak ilcenin nufusu hesaplanır
			for (int j = yeni_mahalle_sayisi - 1; j >= 0; j--) { // mevcut ilceden yeni ilceni mahalleleri cıkartılır
				ilce.mahalleler.remove(j);
			}
			ilce.nufus -= yeni_nufus; // mevcut ilcenin nufusu güncellenir(yeni ilceye mahalleler eklendigi için)
			String ilce_adi = faker.address().cityName(); // yeni ilce adı alınır faker ile
			Ilce yeni_ilce = new Ilce(ilce_adi, yeni_nufus);
			yeni_ilce.mahalleler = yeni_mahalleler; // yeni ilceye yeni mahalleler verilir
			yeni_ilceler.add(yeni_ilce); // yeni ilce ArrayListe eklenir
			return yeni_ilceler;
		} else { // mahalle sayisi 1'dir, o yüzden kisiler ikiye bölünmelidir
			return basamakKontroluKisi(ilce);
		}
	}

	// Kısacası 1 sehrin 1 ilcesi ve 1 mahallesi olma durumudur, kisi sayısı yarıya
	// bölünür
	public static ArrayList<Ilce> basamakKontroluKisi(Ilce ilce) {
		Mahalle mevcut_mahalle = ilce.mahalleler.get(0);
		int yeni_kisi_sayisi = mevcut_mahalle.kisiler.size() / 2;
		ArrayList<Kisi> yeni_kisiler = new ArrayList<>();
		ArrayList<Ilce> yeni_ilceler = new ArrayList<>();
		for (int j = 0; j < yeni_kisi_sayisi; j++) { // mevcut mahallenin kisilerinin yarısı alınır ve yeni_kisiler
														// listesine eklenir
			yeni_kisiler.add(mevcut_mahalle.kisiler.get(j));
		}
		int yeni_nufus = yeni_kisiler.size(); // yeni mahallenin yeni ilcenin nufusu hesaplanir
		for (int j = yeni_kisi_sayisi - 1; j >= 0; j--) {// mevcut mahalleden yeni mahallenin kisileri cikartilir
			mevcut_mahalle.kisiler.remove(j);
		}
		mevcut_mahalle.nufus -= yeni_nufus;// mevcut mahallenin nufusu guncellenir
		ilce.nufus -= yeni_nufus;// mevcut ilcenin nufusu guncellenir
		ArrayList<Mahalle> yeni_mahalleler = new ArrayList<>();
		String mahalle_adi = faker.address().streetName();
		Mahalle yeni_mahalle = new Mahalle(mahalle_adi, yeni_nufus); // yeni bir mahalle olusturulur
		yeni_mahalle.kisiler = yeni_kisiler; // yeni mahalleye yenikisiler eklenir
		yeni_mahalleler.add(yeni_mahalle); // yeni mahalle yeni mahalleler listesine eklenir
		String ilce_adi = faker.address().cityName();
		Ilce yeni_ilce = new Ilce(ilce_adi, yeni_nufus); // yeni ilce olusturulur
		yeni_ilce.mahalleler = yeni_mahalleler; // yeni ilceye yeni mahalleler eklenir
		yeni_ilceler.add(yeni_ilce); // yeni ilcelere yeni ilce eklenir
		return yeni_ilceler;
	}

	// sehirlerin nufuslari 4 basamakli mi kontrolu yapar for dongusu ile
	public static void nufusBasamakSayisiKontrolu(ArrayList<Sehir> sehirler) {
		ArrayList<Sehir> yeniSehirler = new ArrayList<>();// yeni eklenecek sehirleri ıcınde tutar dongu ile tum
															// sehirler kontrol ettikten sonra yeni sehirleri eklemek
															// icin
		for (int s = 0; s < sehirler.size(); s++) {
			if (sehirler.get(s).nufus > 999) {
				ArrayList<Ilce> yeni_ilceler = new ArrayList<>(); // yeni sehire eklenecek yeni ilceleri tutmak icin
																	// liste olusturulur
				yeni_ilceler = basamakKontroluIlce(sehirler.get(s));// yeni sehrin ilcelerini almak icin sehrin
																	// ilcelerini bolen fonksiyonu cagırır.
				int yeni_sehir_nufusu = yeni_ilceler.size() * yeni_ilceler.get(0).nufus;// ilcelerden birinin nufusu ile
																						// ilce sayısını carparak yeni
																						// sehrin nufusu hesaplanir
				sehirler.get(s).nufus -= yeni_sehir_nufusu; // mevcut sehrin nufusu guncellenir
				String sehir_ad = faker.address().city();
				Sehir yeni_sehir = new Sehir(sehir_ad, yeni_sehir_nufusu);
				yeni_sehir.ilceler = yeni_ilceler; // yeni sehre alınan ilceler verilir
				yeniSehirler.add(yeni_sehir); // yeni sehir yeniSehirler listesine eklenir
			}
		}
		sehirler.addAll(yeniSehirler);// yeniSehirler var olan sehirlerin sonuna eklenir.
	}

}
